### PR TITLE
GSoC 2019: Integrates Test Launcher with Mason test

### PR DESCRIPTION
**Pull request by krishnadey30**
This PR integrates the UnitTest Launcher with `mason test`. This is one step towards resolving https://github.com/chapel-lang/chapel/issues/13668.

The approach followed is explained here and was discussed [here](https://github.com/chapel-lang/chapel/issues/13668#issuecomment-523591749):
- If in a mason project, look in `test/` for tests to run
  - If test does not return expected `UnitTest` output, rely on the `exit code`
  - If test does return expected `UnitTest` output, treat as a `UnitTest` test
- If outside of a mason project, look in the provided `<path>` or `<file>` given.

Also modifies the Launcher specific Tests for testing `mason test`.

### CODE REVIEW

File: tools/mason/MasonHelp.chpl [hunk 1]
```diff
@@ -353,6 +353,8 @@ proc masonTestHelp() {
   writeln("    -h, --help                  Display this message");
   writeln("        --show                  Direct output of tests to stdout");
   writeln("        --no-run                Compile tests without running them");
+  writeln("        --keep-binary           Doesn't delete the binaries after running");
+  writeln("        --subdir                Descend recursively into subdirectories of given Directories");
   writeln("        --parallel              Run tests in parallel(sequential by default)");
   writeln();
```
Comment (ben-albrecht) [inline tools/mason/MasonHelp.chpl]: We need to specify that `mason test` works inside and outside of mason packages.

File: tools/mason/MasonHelp.chpl [hunk 2]
```diff
@@ -353,6 +353,8 @@ proc masonTestHelp() {
   writeln("    -h, --help                  Display this message");
```
[the remark this answers is not in the document]
Comment (krishnadey30) [reply tools/mason/MasonHelp.chpl]: I have updated it. Thanks

File: tools/mason/MasonHelp.chpl [hunk 3]
```diff
@@ -353,6 +353,8 @@ proc masonTestHelp() {
...
   writeln("        --show                  Direct output of tests to stdout");
   writeln("        --no-run                Compile tests without running them");
+  writeln("        --keep-binary           Doesn't delete the binaries after running");
+  writeln("        --subdir                Descend recursively into subdirectories of given Directories");
```
[the remark this answers is not in the document]
Comment (krishnadey30) [reply tools/mason/MasonHelp.chpl]: Sure. I made the changes.

File: tools/mason/MasonTest.chpl [hunk 4]
```diff
@@ -27,6 +27,13 @@ use MasonUpdate;
 use MasonBuild;
 use Path;
 use FileSystem;
+use TestResult;
+use Sys;
+
+var subdir = false;
+var keepExec = false;
+config const setComm: string = "";
```
[the remark this answers is not in the document]
Comment (krishnadey30) [reply tools/mason/MasonTest.chpl]: Any alternative to use them?

File: tools/mason/MasonHelp.chpl (hunk 1 above)
Comment (krishnadey30) [reply tools/mason/MasonHelp.chpl]: I have made some changes. Would be helpful if you can look at them.

**Comment by krishnadey30** [comment]
> Hi @krishnadey30 - nice work putting this together. Testing passed for me. We are very close, below is my feedback categorized by when we should aim to get them completed:
> * Compiling with gasnet and running `mason test` with `CHPL_COMM=gasnet` with the locales test from
>   `/test/library/packages/UnitTest` gave:
> 
> ```
> error: Specify number of locales via -nl <#> or --numLocales=<#>
> --- Results ---
> Test: locales Failed
> ```
> 
> This might be due to incorrectly detecting the runtime comm layer.

I have fixed these.

File: tools/mason/MasonTest.chpl (hunk 4 above)
[the remark this answers is not in the document]
Comment (ben-albrecht) [reply tools/mason/MasonTest.chpl]: Parse from args like other command line args.

File: tools/mason/MasonHelp.chpl [hunk 5]
```diff
@@ -344,15 +344,19 @@ proc masonCompilerHelp() {
...
   writeln("Usage:");
-  writeln("    mason test [options]");
+  writeln("    mason test [options] <path>");
   writeln();
   writeln("Options:");
   writeln("    -h, --help                  Display this message");
   writeln("        --show                  Direct output of tests to stdout");
   writeln("        --no-run                Compile tests without running them");
+  writeln("        --keep-binary           Doesn't delete the binaries after running");
+  writeln("        --recursive             Descend recursively into subdirectories of given Directories");
```
Comment (ben-albrecht) [inline tools/mason/MasonHelp.chpl]:
```suggestion
  writeln("        --recursive             Descend recursively into subdirectories of given directories");
```

File: tools/mason/MasonHelp.chpl [hunk 6]
```diff
@@ -344,15 +344,19 @@ proc masonCompilerHelp() {
  }
 
 proc masonTestHelp() {
-  writeln("Run test files located within target/debug/test");
+  writeln("mason test works inside and outside of mason packages.");
+  writeln("Inside a mason package run test files located within target/debug/test");
```
Comment (ben-albrecht) [inline tools/mason/MasonHelp.chpl]:
```suggestion
  writeln("Inside a mason package: run test files found in test/");
```

File: tools/mason/MasonHelp.chpl [hunk 7]
```diff
@@ -344,15 +344,19 @@ proc masonCompilerHelp() {
...
+  writeln("Outside of a mason package runs all the files in the provided path or the current directory.");
```
Comment (ben-albrecht) [inline tools/mason/MasonHelp.chpl]:
```suggestion
  writeln("Outside of a mason package: run test files found in the provided path (defaults to '.').");
```

Comment (krishnadey30) [reply tools/mason/MasonHelp.chpl]: Thanks for the feedback. I have made the changes.

File: tools/mason/MasonHelp.chpl (hunk 6 above)
Comment (krishnadey30) [reply tools/mason/MasonHelp.chpl]: Updated, Thanks

File: tools/mason/MasonHelp.chpl (hunk 5 above)
Comment (krishnadey30) [reply tools/mason/MasonHelp.chpl]: Updated, Thanks

File: tools/mason/MasonHelp.chpl [hunk 8]
```diff
@@ -344,17 +344,22 @@ proc masonCompilerHelp() {
...
   writeln();
   writeln("Options:");
   writeln("    -h, --help                  Display this message");
   writeln("        --show                  Direct output of tests to stdout");
   writeln("        --no-run                Compile tests without running them");
+  writeln("        --keep-binary           Doesn't delete the binaries after running");
+  writeln("        --recursive             Descend recursively into subdirectories of given directories");
   writeln("        --parallel              Run tests in parallel(sequential by default)");
   writeln("        --[no]-update           [Do not] update the mason-registry when testing");
+  writeln("        --setComm               Set the CHPL_COMM value for running the tests,  e.g. none, gasnet, ugni");
```
Comment (ben-albrecht) [inline tools/mason/MasonHelp.chpl]: I think this should include a 2nd line specifying the default, e.g.

```
Defaults to current value of CHPL_COMM.
```

**Pull request merged**